### DISCUSSION TITLE
Clarification for Ack Sequence

### DISCUSSION THREAD

**Discussion by kohlisid** [nats-io/nats-server#5737] [General]
We have seen this on one of our stream consumers, we are using WorkQueue with DiscardNew
```
Configuration:

                    Name: **
               Pull Mode: true
          Filter Subject: ***
          Deliver Policy: All
              Ack Policy: Explicit
                Ack Wait: 1m0s
           Replay Policy: Instant
         Max Ack Pending: 25,000
       Max Waiting Pulls: 512

Cluster Information:

                    Name: default
                  Leader: isbsvc-default-js-1
                 Replica: isbsvc-default-js-0, current, seen 4ms ago
                 Replica: isbsvc-default-js-4, current, seen 4ms ago

State:

  Last Delivered Message: Consumer sequence: 36,874,712 Stream sequence: 36,874,472 Last delivery: 6ms ago
    Acknowledgment Floor: Consumer sequence: 36,874,712 Stream sequence: 36,874,472 Last Ack: 585µs ago
        Outstanding Acks: 0 out of maximum 25,000
    Redelivered Messages: 0
    Unprocessed Messages: 92
           Waiting Pulls: 0 of maximum 512
```
Can someone help me to understand these numbers better and how can we ensure that there are no messages getting dropped by the stream.

From the docs it says that

```
The two number are not directly related: the Stream sequence number is the pointer to the exact message, while the Consumer sequence number is an ever-increasing counter for consumer actions.
```
https://docs.nats.io/running-a-nats-service/nats_admin/jetstream_admin/consumers#stream-vs-consumer-sequence-numbers

But what do they signify, would like some more clarity

**Comment by derekcollison** [comment]
Stream can be considered unique and stable index to the message. Consumer sequences start at 1 and always montonically increase by one for every time the consumer state engine in the server delivers a message. Each message received in your app has both of those plus what the system feels is the number of deliveries for that message. Meaning if the message is failed to be ack'd (or nak'd or in progress) before ack wait, we will deliver same message (same stream sequence) again with a different consumer sequence and NumDelivered = 2.

**Comment by kohlisid** [comment]
In from the above logs we can see that the redelivered messages are 0, unprocessed 92 and no outstanding acks as well
But still the consumer sequence and stream sequence is skewed 
```
 Last Delivered Message: Consumer sequence: 36,874,712 Stream sequence: 36,874,472 Last delivery: 6ms ago
    Acknowledgment Floor: Consumer sequence: 36,874,712 Stream sequence: 36,874,472 Last Ack: 585µs ago
        Outstanding Acks: 0 out of maximum 25,000
    Redelivered Messages: 0
```
@derekcollison

**Reply by derekcollison** [reply]
That reporting is tied to server lifetime but the consumer state is persisted. So possible to see 0 redelivered but they could have happened prior to a server restart.

**Comment by ripienaar** [comment]
With a filter subject matching only some of the stream messages these sequences can’t match. 

Your consumer info suggests you have a filter but not clear if it’s matching all or not?

irs best to just work on the assumption these are not related and don’t increment in lock step.
